### DISCUSSION TITLE
FATAL: too many connections for role

### DISCUSSION THREAD

**Discussion by rserafim** [prisma/prisma#14204] [Q&A] [answered]
### Bug description

ERROR	PrismaClientInitializationError: Error querying the database: db error: FATAL: too many connections for role "name-bd"

### How to reproduce

It happens randomly. In my case it's the Google bot indexing the pages. It ran for several days without errors. And today presented.

![image](https://user-images.githubusercontent.com/48859044/177888048-fd981cbd-3f86-44b5-97aa-11e843fd2988.png)

### Prisma information

```
import { PrismaClient } from '@prisma/client'

declare global {
  // allow global `var` declarations
  // eslint-disable-next-line no-var
  var prisma: PrismaClient | undefined
}

export const prisma =
  global.prisma ||
  new PrismaClient({
    log: ['query'],
  })

if (process.env.NODE_ENV !== 'production') global.prisma = prisma
```

QUERY
```
const result = await prisma.$queryRaw<Company[]>
    SELECT
      c.id, c.name, c.phones, c.slug, c.information, c.street, c.number,
      c.neighborhood, c.whatsapp, c.email, c.site, c.facebook, c.instagram,
      c.zipcode, c.likes, c.services, c.created_at, c.updated_at,
      c.slugneighborhood, c.state_id, c.city_id, c.category_id,
      ct.name as city_name, ct.slug as city_slug, st.name as state_name,
      st.slug as state_slug, st.uf as state_uf, cat.name as cat_name, cat.slug as cat_slug,
      catf.name as catf_name, catf.slug as catf_slug
    FROM zapia.companies as c
    INNER JOIN zapia.cities as ct on c.city_id = ct.id
    INNER JOIN zapia.states as st on c.state_id =st.id
    INNER JOIN zapia.categories as cat on c.category_id = cat.id
    INNER JOIN zapia.categories_fathers as catf on cat.categories_father_id = catf.id
    WHERE c.id = ${idNumber}
    AND c.visible = true
    LIMIT 1
 ```

### Environment & setup

- Database: postgres
- Vercel : Plan Pro
- "next": "12.1.0",

### Prisma Version

```
"prisma": "^3.15.1",
 "@prisma/client": "^3.15.1",
```

**Comment by TasinIshmam** [comment, marked as answer]
Hey @rserafim  👋  

Could you clarify if this happens in your production environment only? Does it also happen in your development environment? 


It's very likely that you're exhausting your connection limit to the database when there is excessive load, due to too many function instances being spun up and trying to connect to the database. Using an [external connection pooler](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#external-connection-poolers) should solve your problem.

**Reply by rserafim** [reply]
Hello @TasinIshmam , thanks for the reception and quick response.

Only in production this error.

As I understand it, in my Deploying to Vercel Serverless Functions case, just add a variable PRISMA_GENERATE_DATAPROXY = true

Right?
https://www.prisma.io/docs/data-platform/data-proxy#deploying-to-vercel-serverless-functions
https://newreleases.io/project/github/prisma/prisma/release/3.15.2

**Reply by TasinIshmam** [reply]
> As I understand it, in my Deploying to Vercel Serverless Functions case, just add a variable PRISMA_GENERATE_DATAPROXY = true

You have to follow all the steps mentioned in the [prisma docs](https://www.prisma.io/docs/data-platform/data-proxy#using-the-data-proxy-in-a-prisma-application).  Notably, you need to sign up to the [Prisma Data Platform](https://cloud.prisma.io/) first. 

We have a tutorial about using Data proxy in a Next.js app deployed to Vercel [on youtube](https://www.youtube.com/watch?v=iyGZ8JFPgoY ). Note that this was recorded a while ago when the Data proxy was still in early access. So in case of any discrepancies with the docs, follow what is mentioned in the docs.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
